cpp/src/groupby/hash/compute_aggregations.cuh

```diff
@@ -0,0 +1,175 @@
+                                    (shmem_offsets_size(flattened_values.num_columns()) * 2);
+  auto const has_dictionary_request = std::any_of(
+    requests.begin(), requests.end(), [](cudf::groupby::aggregation_request const& request) {
+      return cudf::is_dictionary(request.values.type());
+    });
+  auto const is_shared_memory_compatible = !has_dictionary_request and has_sufficient_shmem;
+
+  // Performs naive global memory aggregations when the workload is not compatible with shared
+  // memory, such as when aggregating dictionary columns or when there is insufficient dynamic
+  // shared memory for shared memory aggregations.
+  if (!is_shared_memory_compatible) {
+    return compute_global_memory_aggs(num_rows,
+                                      skip_rows_with_nulls,
+                                      row_bitmask,
+                                      flattened_values,
+                                      d_agg_kinds.data(),
+                                      agg_kinds,
+                                      global_set,
+                                      aggs,
+                                      sparse_results,
+                                      stream);
+  }
+
+  // 'populated_keys' contains inserted row_indices (keys) of global hash set
+  rmm::device_uvector<cudf::size_type> populated_keys(num_rows, stream);
+  // 'local_mapping_index' maps from the global row index of the input table to its block-wise rank
+  rmm::device_uvector<cudf::size_type> local_mapping_index(num_rows, stream);
+  // 'global_mapping_index' maps from the block-wise rank to the row index of global aggregate table
+  rmm::device_uvector<cudf::size_type> global_mapping_index(grid_size * GROUPBY_SHM_MAX_ELEMENTS,
+                                                            stream);
+  rmm::device_uvector<cudf::size_type> block_cardinality(grid_size, stream);
+
+  // Flag indicating whether a global memory aggregation fallback is required or not
+  rmm::device_scalar<cuda::std::atomic_flag> needs_global_memory_fallback(stream);
+
+  auto global_set_ref = global_set.ref(cuco::op::insert_and_find);
+
+  compute_mapping_indices(grid_size,
+                          num_rows,
+                          global_set_ref,
+                          row_bitmask,
+                          skip_rows_with_nulls,
+                          local_mapping_index.data(),
+                          global_mapping_index.data(),
+                          block_cardinality.data(),
+                          needs_global_memory_fallback.data(),
+                          stream);
+
+  cuda::std::atomic_flag h_needs_fallback;
+  CUDF_CUDA_TRY(cudaMemcpyAsync(&h_needs_fallback,
+                                needs_global_memory_fallback.data(),
+                                sizeof(cuda::std::atomic_flag),
+                                cudaMemcpyDefault,
+                                stream.value()));
```

Comment:
Both the `rmm::device_scalar` and `cudf::detail::device_scalar` have a `value(stream)` member that will do this copy for you.
```suggestion
  cuda::std::atomic_flag h_needs_fallback = needs_global_memory_fallback.value(stream);
```

Reply:
I started by using the `value` member but `atomic_flag` doesn't have a copy ctor thus both `cudf::detail::device_scalar` and `rmm::device_scalar` won't build:
```bash
(rapids) coder ➜ ~ $ build-cudf-cpp -j30 -DBUILD_BENCHMARKS=ON
[2/165] Building CUDA object CMakeFiles/cudf.dir/src/groupby/hash/compute_aggregations.cu.o
FAILED: CMakeFiles/cudf.dir/src/groupby/hash/compute_aggregations.cu.o 
/usr/bin/sccache /usr/local/cuda/bin/nvcc -forward-unknown-to-host-compiler -ccbin=/usr/bin/g++ -DBS_THREAD_POOL_ENABLE_PAUSE=1 -DCUFILE_FOUND -DCURL_STATICLIB -DFMT_HEADER_ONLY=1 -DJITIFY_PRINT_LOG=0 -DKVIKIO_CUDA_FOUND -DKVIKIO_CUFILE_BATCH_API_FOUND -DKVIKIO_CUFILE_FOUND -DKVIKIO_CUFILE_STREAM_API_FOUND -DKVIKIO_LIBCURL_FOUND -DLIBCUDACXX_ENABLE_EXPERIMENTAL_MEMORY_RESOURCE -DRMM_LOGGING_LEVEL=LIBCUDF_LOGGING_LEVEL -DSPDLOG_ACTIVE_LEVEL=SPDLOG_LEVEL_INFO -DSPDLOG_FMT_EXTERNAL -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_CUDA -DTHRUST_DISABLE_ABI_NAMESPACE -DTHRUST_HOST_SYSTEM=THRUST_HOST_SYSTEM_CPP -DTHRUST_IGNORE_ABI_NAMESPACE_ERROR -Dcudf_EXPORTS -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/dlpack-src/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/jitify-src -I/home/coder/cudf/cpp/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/include -I/home/coder/cudf/cpp/src -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/nanoarrow-src/src -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/flatbuffers-src/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/cccl-src/thrust/thrust/cmake/../.. -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/cccl-src/libcudacxx/lib/cmake/libcudacxx/../../../include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/cccl-src/cub/cub/cmake/../.. -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/rmm-src/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/fmt-src/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/spdlog-src/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/nvtx3-src/c/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/bs_thread_pool-src/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/cuco-src/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/kvikio-src/cpp/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/curl-src/include -I/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/nanoarrow-build/src -isystem /usr/local/cuda/targets/x86_64-linux/include -isystem /home/coder/cudf/cpp/build/latest/_deps/nvcomp_proprietary_binary-src/include -t=1 -O3 -DNDEBUG -std=c++17 "--generate-code=arch=compute_70,code=[sm_70]" "--generate-code=arch=compute_75,code=[sm_75]" "--generate-code=arch=compute_80,code=[sm_80]" "--generate-code=arch=compute_86,code=[sm_86]" "--generate-code=arch=compute_90,code=[compute_90,sm_90]" -Xcompiler=-fPIC -Xcompiler=-fvisibility=hidden --expt-extended-lambda --expt-relaxed-constexpr -Werror=all-warnings -Xcompiler=-Wall,-Werror,-Wno-error=deprecated-declarations -diag-suppress=1407 -Xfatbin=-compress-all -MD -MT CMakeFiles/cudf.dir/src/groupby/hash/compute_aggregations.cu.o -MF CMakeFiles/cudf.dir/src/groupby/hash/compute_aggregations.cu.o.d -x cu -c /home/coder/cudf/cpp/src/groupby/hash/compute_aggregations.cu -o CMakeFiles/cudf.dir/src/groupby/hash/compute_aggregations.cu.o
/home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/rmm-src/include/rmm/device_uvector.hpp(327): error: function "cuda::std::__4::atomic_flag::atomic_flag(const cuda::std::__4::atomic_flag &)" (declared at line 642 of /home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/cccl-src/libcudacxx/lib/cmake/libcudacxx/../../../include/cuda/std/atomic) cannot be referenced -- it is a deleted function
      return value;
             ^
          detected during:
            instantiation of "rmm::device_uvector<T>::value_type rmm::device_uvector<T>::element(std::size_t, rmm::cuda_stream_view) const [with T=cuda::std::__4::atomic_flag]" at line 343
            instantiation of "rmm::device_uvector<T>::value_type rmm::device_uvector<T>::front_element(rmm::cuda_stream_view) const [with T=cuda::std::__4::atomic_flag]" at line 164 of /home/coder/cudf/cpp/build/pip/cuda-12.5/release/_deps/rmm-src/include/rmm/device_scalar.hpp
            instantiation of "rmm::device_scalar<T>::value_type rmm::device_scalar<T>::value(rmm::cuda_stream_view) const [with T=cuda::std::__4::atomic_flag]" at line 118 of /home/coder/cudf/cpp/src/groupby/hash/compute_aggregations.cuh
            instantiation of "rmm::device_uvector<cudf::size_type> cudf::groupby::detail::hash::compute_aggregations(int64_t, __nv_bool, const cudf::bitmask_type *, SetType &, cudf::host_span<const cudf::groupby::aggregation_request, 18446744073709551615UL>, cudf::detail::result_cache *, rmm::cuda_stream_view) [with SetType=cudf::groupby::detail::hash::global_set_t]" at line 21 of /home/coder/cudf/cpp/src/groupby/hash/compute_aggregations.cu
```

Reply:
Ah, ok. Good to know. Maybe a comment here then as well.